From 6e56df1a3aa631feac70ea0c4e1db8b614bd445c Mon Sep 17 00:00:00 2001
From: Cool Lee <cool_lee@aspeedtech.com>
Date: Fri, 25 Jul 2025 16:02:55 +0800
Subject: [PATCH] ram: sdram_ast2700: Correct vga node detect rule

Original rule is checking OTP only.
The truth is vga node limited by efuse chip id and the OTP used to
disable.

Signed-off-by: Cool Lee <cool_lee@aspeedtech.com>
Change-Id: Ie227e07bd7ff8e4ebbd58fef13784e12024c0188

(cherry picked from commit 3562546c7c937985f2f0470f68dfcd2c0be2c6ac)
Signed-off-by: Vasanthakumar K <vasanthakumark@ami.com>
---
 drivers/ram/aspeed/sdram_ast2700.c | 34 +++++++-----------------------
 1 file changed, 8 insertions(+), 26 deletions(-)

diff --git a/drivers/ram/aspeed/sdram_ast2700.c b/drivers/ram/aspeed/sdram_ast2700.c
index 4c802403ef..33422d55c0 100644
--- a/drivers/ram/aspeed/sdram_ast2700.c
+++ b/drivers/ram/aspeed/sdram_ast2700.c
@@ -18,6 +18,7 @@
 #include <asm/arch/sdram_ast2700.h>
 #include <asm/arch/scu_ast2700.h>
 #include <asm/arch/platform.h>
+#include <aspeed_vga.h>
 
 #define DRAMC_UNLOCK_KEY		0x1688a8a8
 #define DRAMC_VIDEO_UNLOCK_KEY		0x00440003
@@ -685,45 +686,26 @@ static void sdramc_aes_enable(struct sdramc *sdramc)
 #define DRAM_SIZE_DEF	3
 static size_t ast2700_sdrammc_get_vga_mem_size(struct sdramc *sdramc)
 {
-	struct ast2700_scu0 *scu;
 	struct sdramc_regs *regs = sdramc->regs;
-	int nodeoffset;
 	u32 vga_ram_size[] = {
 		0x2000000, // 32MB
 		0x4000000, // 64MB
 		};
 	int vga_sz_sel;
-	ofnode node;
-	int dual = 0;
+	int vga_cnt;
 
 	vga_sz_sel = readl(&regs->gfmcfg) & 0x1;
 
-	/* find the offset of compatible node */
-	nodeoffset = fdt_node_offset_by_compatible(gd->fdt_blob, -1,
-						   "aspeed,ast2700-scu0");
-	if (nodeoffset < 0) {
-		printf("%s: failed to get aspeed,ast2700-scu0\n", __func__);
-		return -ENODEV;
-	}
+	vga_cnt = ast_vga_get_nodes(NULL, NULL);
 
-	/* get ast2700-scu0 node */
-	node = offset_to_ofnode(nodeoffset);
-
-	scu = (struct ast2700_scu0 *)ofnode_get_addr(node);
-
-	if (scu->pci0_misc[28] & BIT(0)) {
-		if (!IS_ENABLED(CONFIG_SPL_BUILD))
-			printf("VGA0:%dMiB, ", vga_ram_size[vga_sz_sel] / SZ_1M);
-		dual++;
-	}
-
-	if (scu->pci1_misc[28] & BIT(0)) {
-		if (!IS_ENABLED(CONFIG_SPL_BUILD))
+	if (!IS_ENABLED(CONFIG_SPL_BUILD)) {
+		if (vga_cnt > 1)
 			printf("VGA1:%dMiB, ", vga_ram_size[vga_sz_sel] / SZ_1M);
-		dual++;
+		if (vga_cnt > 0)
+			printf("VGA0:%dMiB, ", vga_ram_size[vga_sz_sel] / SZ_1M);
 	}
 
-	return vga_ram_size[vga_sz_sel] * dual;
+	return vga_ram_size[vga_sz_sel] * vga_cnt;
 }
 
 static int sdramc_ecc_enable(struct sdramc *sdramc)
-- 
2.34.1

